Firmware/MVP/battsci.cpp

```diff
@@ -86,6 +98,169 @@ uint8_t BATTSCI_calculateChecksum( uint8_t frameSum )
+
+    @param      evalSoC     Integer value of SoC in 0.1% increments beginning at 20%. 0 = 20.0%
+    @return                 This function does not return anything.
+  */
+
+  // battTempBATTSCI will be displayed if it won't affect MCM behaviour, as determined by SoC
+  // 0x26 = 08 Deg C
+  // 0x30 = 18 Deg C
+  // 0x31 = 19 Deg C
+  int8_t battTempBATTSCI = temperature_battery_getLatest() + 30;                      //T_MCM = T_actual + 30
+
+  if (evalSoC >= 200)             // 200 = 40% SoC or higher
+  {
+	if (battTempBATTSCI > 0x33) {
+		temperature_Byte = battTempBATTSCI;
+	} else temperature_Byte = 0x33;      // Set temperature to +21 deg C to allow max Assist in 2nd and 3rd
+  } else {
+	if (battTempBATTSCI < 0x30) {
+		temperature_Byte = battTempBATTSCI;
+	} else temperature_Byte = 0x30;      // Set temperature to +18 deg C to reduce max Assist in 2nd and 3rd
+  }
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void BATTSCI_calculateSoC()
+{
+  /**
+    BATTSCI_calculateSoC calculates what SoC we want to send the MCM, in order to govern the MCM's behaviour.
+    The MCM changes how the car uses assist and regen depending on the SoC number we send to it.
+
+    @return                 This function does not return anything.
+  */
+  LiBCM_SoC = SoC_getBatteryStateNow_percent();
+
+  if (LiBCM_SoC >= 95) {
+    calculatedSoC = 620;          // No Regen Allowed, set MCM SoC high to enforce a cooldown period (while SoC drops) before regen is allowed again.
+    oldCalculatedSoC = 620;       // Make sure SoC doesn't immediately spike back down
+  } else if (LiBCM_SoC >= 90) {
+    calculatedSoC = map(LiBCM_SoC, 90, 94, 600, 619);   // No Regen Allowed
+  } else if (LiBCM_SoC >= 60) {
+    calculatedSoC = map(LiBCM_SoC, 60, 89, 501, 599);   // No BG Regen Allowed 84 = 582
+  } else if (LiBCM_SoC >= 30) {
+    calculatedSoC = map(LiBCM_SoC, 30, 59, 151, 500);   // BG Regen Allowed
+  } else if (LiBCM_SoC >= 5) {
+    calculatedSoC = map(LiBCM_SoC, 5, 29, 1, 150);      // BG Regen Allowed, SoC very low.
+  } else {
+    // No Assist Allowed
+    calculatedSoC = 0;
+    oldCalculatedSoC = 0;         // Make sure SoC doesn't immediately spike back up
+    BATTSCI_evaluateSoCBytes(calculatedSoC);
+  }
+
+  // On first run we set oldCalculatedSoC and calculatedSoC to the same value.
+  if (initializeSoC) {
+    oldCalculatedSoC = calculatedSoC;
+    initializeSoC = false;
+    BATTSCI_evaluateSoCBytes(calculatedSoC);
+  }
+
+  static int16_t packMilliAmps = 0;
+  packMilliAmps = adc_getLatestBatteryCurrent_amps();
+
+  // Modify calculatedSoC in place to be an increment of oldCalculatedSoC
+  // packMilliAmps is being checked so that we only increment SoC if we have current < -0.01 Amps or decremented if we have > +0.01 Amps
+  // This should prevent SoC being changed during Auto Stop.
+
+  if (calculatedSoC > oldCalculatedSoC) {
```

Comment:
This is blending the spoofed SoC we send to MCM to make the changes less abrupt, and is related to me trying to solve the SoC gauge freeze issue.  One of the most important functions this does is it stops spoofed SoC from going up when assist is being used and it stops spoofed SoC from going down when regening.

Discovering Peter's google drive doc of a BATTSCI dump has informed me that this can be modified to allow larger SoC change steps between updates (SoC sent to MCM is only updated every 16 or 17 frames by stock BCM) but I haven't gotten to making those changes yet.

Reply:
What would cause the spoofed SoC to increase during assist (or decrease during regen)?  When the key is on, the actual SoC LiBCM calculates is purely a function of current into/out-of the pack.  Why would the spoofed SoC be any different?  Maybe when we hang out around 70% SoC?  I guess I don't understand the entire concept of having a spoofed SoC that's anything more than a remapped version of the actual SoC.